### PR TITLE
Improve audio featurizer and add shift augmentor for DS2.

### LINKED ISSUE #113: Improve audio featurizer and add shift augmentor for DS2.
Improve audio featurizer by adding e.g. resampling, db_normalization, and random shift, as suggested in speech_dl codes.

### PR DESCRIPTION
**Pull request by xinghai-sun**
resolve #113

1. Improve audio featurizer (resample, db_normalize, and random shift), as suggested in the speech_dl codes.
2. Add shift augmentor.
3. Update default arguments to be the current best seggestion.
4. Add checkpoints with pass id.

Training experiment is in progress, and its results will be pasted here ASAP.

### CODE REVIEW

File: deep_speech_2/data_utils/audio.py [hunk 1]
```diff
@@ -67,6 +67,54 @@ def from_file(cls, file):
         return cls(samples, sample_rate)
 
     @classmethod
+    def slice_from_file(cls, file, start=None, end=None):
```
Comment (xinghai-sun) [inline deep_speech_2/data_utils/audio.py:70]: @reviewers:
No different for `slice_from_file` and `make_silence`. Only re-order them.

File: deep_speech_2/data_utils/audio.py [hunk 2]
```diff
@@ -366,6 +366,31 @@ def pad_silence(self, duration, sides='both'):
...
+        keep the duration unchanged.
+
+        Note that this is an in-place transformation.
+
+        :param shift_ms: Shift time in millseconds. If positive, shift with
+                         time advance; if negative; shift with time delay.
+        :type shift_ms: float
+        :raises ValueError: If shift_ms is longer than audio duration.
+        """
+        if shift_ms / 1000.0 > self.duration:
```
Comment (pkuyym) [inline deep_speech_2/data_utils/audio.py]: Should be ```abs(shift_ms)``` ?

File: deep_speech_2/data_utils/featurizer/audio_featurizer.py [hunk 3]
```diff
@@ -24,26 +24,64 @@ class AudioFeaturizer(object):
                      corresponding to frequencies between [0, max_freq] are
                      returned.
     :types max_freq: None|float
+    :param target_sample_rate: Audio are resampled (if upsampling or
+                               downsampling is allowed) to this before
+                               extracting spectrogram features.
+    :type target_sample_rate: float
+    :param use_dB_normalization: Whether to normalize the audio to a certain
+                                 decibels before extracting the features.
```
Comment (pkuyym) [inline deep_speech_2/data_utils/featurizer/audio_featurizer.py:32]: Better to change **decibels** to **dB** for consistency

File: deep_speech_2/data_utils/featurizer/audio_featurizer.py [hunk 4]
```diff
@@ -24,26 +24,64 @@ class AudioFeaturizer(object):
...
+        # upsampling or downsampling
+        if ((audio_segment.sample_rate > self._target_sample_rate and
+             allow_downsampling) or
+            (audio_segment.sample_rate < self._target_sample_rate and
+             allow_upsampling)):
+            audio_segment.resample(self._target_sample_rate)
+        if audio_segment.sample_rate != self._target_sample_rate:
+            raise ValueError("Audio sample rate is not supported. "
+                             "Turn allow_downsampling or allow up_sampling on.")
+        # decibel normalization
```
Comment (pkuyym) [inline deep_speech_2/data_utils/featurizer/audio_featurizer.py:81]: dB better ?

File: deep_speech_2/data_utils/featurizer/speech_featurizer.py [hunk 5]
```diff
@@ -29,16 +29,34 @@ class SpeechFeaturizer(object):
...
                      returned.
     :types max_freq: None|float
+    :param target_sample_rate: Speech are resampled (if upsampling or
+                               downsampling is allowed) to this before
+                               extracting spectrogram features.
+    :type target_sample_rate: float
+    :param use_dB_normalization: Whether to normalize the audio to a certain
+                                 decibels before extracting the features.
+    :type use_dB_normalization: bool
+    :param target_dB: Target audio decibels for normalization.
```
Comment (pkuyym) [inline deep_speech_2/data_utils/featurizer/speech_featurizer.py:39]: dB better?

**Review by pkuyym — CHANGES_REQUESTED** [review]
Almost LGTM.

File: deep_speech_2/data_utils/audio.py (hunk 2 above)
Comment (xinghai-sun) [reply deep_speech_2/data_utils/audio.py]: Done.

Comment (xinghai-sun) [reply deep_speech_2/data_utils/audio.py]: Done.

File: deep_speech_2/data_utils/featurizer/audio_featurizer.py (hunk 3 above)
Comment (xinghai-sun) [reply deep_speech_2/data_utils/featurizer/audio_featurizer.py:32]: For comments, full name `decibels` is used for clarity, while in arguments a short name of `dB` is used instead.
I think it makes sense?

File: deep_speech_2/data_utils/featurizer/audio_featurizer.py (hunk 4 above)
Comment (xinghai-sun) [reply deep_speech_2/data_utils/featurizer/audio_featurizer.py:81]: For comments, full name `decibels` is used for clarity, while in arguments a short name of `dB` is used instead.
I think it makes sense?

File: deep_speech_2/data_utils/featurizer/speech_featurizer.py (hunk 5 above)
Comment (xinghai-sun) [reply deep_speech_2/data_utils/featurizer/speech_featurizer.py:39]: For comments, full name `decibels` is used for clarity, while in arguments a short name of `dB` is used instead.
I think it makes sense?

**Review by pkuyym — APPROVED** [review]
Great. LGTM.

**Comment by xinghai-sun** [comment]
The experimental results for comparing old and new featurizer are as follows. We have a better convergence with the new featurizer.

compare_old_new_data

**Pull request merged**